rust/src/mask/masking.rs

```diff
@@ -0,0 +1,529 @@
+use anyhow::{anyhow, Context};
+use rand::{RngCore, SeedableRng};
+use std::{
+    convert::TryInto,
+    iter::{self, Iterator},
+    ops::Range,
+};
+
+use num::{
+    bigint::{BigInt, BigUint, ToBigInt},
+    clamp,
+    rational::Ratio,
+    traits::Zero,
+};
+use rand_chacha::ChaCha20Rng;
+
+use crate::{
+    mask::{config::MASK_CONFIG_BUFFER_LEN, MaskConfig, MaskSeed, Model},
+    message::{utils::range, DecodeError, FromBytes, ToBytes},
+};
+
+use thiserror::Error;
+
+#[derive(Debug, Hash, PartialEq, Eq, Clone)]
+pub struct MaskObject {
+    data: Vec<BigUint>,
+    config: MaskConfig,
+}
+
+impl MaskObject {
+    pub fn new(config: MaskConfig, data: Vec<BigUint>) -> Self {
+        Self { data, config }
+    }
+
+    pub fn new_empty(config: MaskConfig) -> Self {
+        Self {
+            data: vec![],
+            config,
+        }
+    }
+
+    pub fn new_checked(config: MaskConfig, data: Vec<BigUint>) -> Result<Self, InvalidMaskObject> {
+        let obj = Self::new(config, data);
+        if obj.is_valid() {
+            Ok(obj)
+        } else {
+            Err(InvalidMaskObject)
+        }
+    }
+
+    pub fn is_valid(&self) -> bool {
+        self.data.iter().all(|i| i < &self.config.order())
+    }
```

Comment:
this is a more general rust question: `self.config.order()` creates a new `BigUint` when called, does this mean that a new `BigUint` is created every time `is_valid()` and therefore `order()` is called, or does the compiler place the object somewhere in memory and only returns a reference to it in subsequent calls? i would assume the former since the object gets dropped at the end of `is_valid()`, but i'm not sure. this line of thought was the reason i implemented it before as `MaskConfigs` and `MaskConfig`.

Reply:
No a new BigUint is created everytime. I think this cost is rather marginal though, but if it proves to be an issue, we can always make the `MaskConfig.order` field an `Rc<BigUint>` in the future.